### PR TITLE
8247514: Improve clhsdb 'findpc' ability to determine what an address points to by improving PointerFinder and PointerLocation classes

### PR DESCRIPTION
**Pull request by plummercj**
See the bug for most details. A few notes here about some implementation details:

In the `PointerLocation` class, I added more consistency w.r.t. whether or not a newline is printed. It used to for some address types, but not others. Now it always does. And if you see a comment something like the following:

`         getTLAB().printOn(tty); // includes "\n" `

That's just clarifying whether or not the `printOn()` method called will include the newline. Some do and some don't, and knowing what the various `printOn()` methods do makes getting the proper inclusion of the newline easier to understand.

I added `verbose` and `printAddress` boolean arguments to `PointerLocation.printOn()`. Currently they are always `true`. The false arguments will be used when I complete [JDK-8250801](https://bugs.openjdk.java.net/browse/JDK-8250801), which will use `PointerFinder/Location` to show what each register points to.

The CR mentions that the main motivation for this work is for eventual replacement of the old clhsdb `whatis` command, which was implemented in javascript. It used to resolve DSO symbols, whereas `findpc` did not. The `whatis` code did this with the following:

```
  var dso = loadObjectContainingPC(addr);
  if (dso == null) {
    return ptrLoc.toString();
  }
  var sym = dso.closestSymbolToPC(addr);
  if (sym != null) {
    return sym.name + '+' + sym.offset;
  }
```
And now you'll see something similar in the PointerFinder code:

```
        loc.loadObject = cdbg.loadObjectContainingPC(a);
        if (loc.loadObject != null) {
            loc.nativeSymbol = loc.loadObject.closestSymbolToPC(a);
            return loc;
        } 
```
Note that now that `findpc` does everything that `whatis` used to (and more), we don't really need to add a java version of `whatis`, but I'll probably do so anyway just help out people who are used to using the `whatis` command. That will be done using [JDK-8244670](https://bugs.openjdk.java.net/browse/JDK-8244670)

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8247514](https://bugs.openjdk.java.net/browse/JDK-8247514): Improve clhsdb 'findpc' ability to determine what an address points to by improving PointerFinder and PointerLocation classes

### Reviewers
 * [Yasumasa Suenaga](https://openjdk.java.net/census#ysuenaga) (@YaSuenag - **Reviewer**)
 * [Kevin Walls](https://openjdk.java.net/census#kevinw) (@kevinjwalls - Committer)

### Download
`$ git fetch https://git.openjdk.java.net/jdk pull/2111/head:pull/2111`
`$ git checkout pull/2111`

### CODE REVIEW

**Comment by plummercj** [comment]
Ping!

**Comment by plummercj** [comment]
Ping again.

**Review by YaSuenag — APPROVED** [review]
LGTM

File: src/jdk.hotspot.agent/share/classes/sun/jvm/hotspot/utilities/PointerLocation.java [hunk 1]
```diff
@@ -186,24 +213,66 @@ public String toString() {
...
+      metadata.printValueOn(tty); // does not include "\n"
+      tty.println();
+    } else if (isCtype()) {
+      tty.println("Is of type " + ctype.getName());
+    } else if (isInJavaStack()) {
+        if (verbose) {
+            tty.format("In java stack [%s,%s,%s] for thread %s:\n   ",
+                       stackThread.getStackBase(), stackThread.lastSPDbg(),
+                       stackThread.getStackBase().addOffsetTo(-stackThread.getStackSize()),
+                       stackThread);
```
Comment (kevinjwalls) [inline src/jdk.hotspot.agent/share/classes/sun/jvm/hotspot/utilities/PointerLocation.java:247]: When we print a JavaThread, in the verbose block,
the final argument to tty.format in line 247, I wonder what that prints?

We then call printThreadInfoOn() which will first print the quoted thread name,
so maybe we don't need that item.
Or maybe we want the JavaThread.toString()?

**Review by kevinjwalls — APPROVED** [review]
Looks good, thanks.
(Comment in PointerLocation.java, treat as you see fit.)

File: src/jdk.hotspot.agent/share/classes/sun/jvm/hotspot/utilities/PointerLocation.java (hunk 1 above)
Comment (plummercj) [reply src/jdk.hotspot.agent/share/classes/sun/jvm/hotspot/utilities/PointerLocation.java:247]: `stackThread.toString()` ends up in `VMObject.toString()`:

```
  public String toString() {
    return getClass().getName() + "@" + addr;
  }
```
And here's an example output:
```
hsdb> + findpc 0x0000152f45df6000
Address 0x0000152f45df6000: In java stack [0x0000152f45df8000,0x0000152f45df6580,0x0000152f45cf7000] for thread sun.jvm.hotspot.runtime.JavaThread@0x0000152f3c026f70:
   "main" #1 prio=5 tid=0x0000152f3c026f70 nid=0x308e waiting on condition [0x0000152f45df6000]
   java.lang.Thread.State: TIMED_WAITING (sleeping)
   JavaThread state: _thread_blocked
```
So I think the `stackThread` argument is doing what was intended, and there is no duplication in the output.

Comment (kevinjwalls) [reply src/jdk.hotspot.agent/share/classes/sun/jvm/hotspot/utilities/PointerLocation.java:247]: Great, thanks.

**Comment by plummercj** [comment]
@YaSuenag and @kevinjwalls I had to make a minor fix to the test. Can you please review it. The issued turned up when I ran some higher test tiers, one of which enabled CDS with some tracing and the other enabled `-Xcheck:jni`, which produced output due to [JDK-8261607](https://bugs.openjdk.java.net/browse/JDK-8261607). Both caused extra output  that resulted in improperly parsing the `examine` output and not actually finding that address that it produced. This was because there were lines of output before even issuing the `examine` command that matched the pattern being looked for. I made the pattern more specific by including the tid of the thread. I also cleaned up the comments around that code a bit. thanks.

**Review by kevinjwalls — APPROVED** [review]
Yes, joy of text processing.  Looks good.

**Review by YaSuenag — APPROVED** [review]
LGTM

We may be able to use regex to collect any addresses from jstack output, but I'm not sure it makes the test code simpler...

**Pull request closed without merging**